Refresh COMMAND, add topics: command-args, key-specs (#1697)
redis/redis-doc: commands/command.md

Returns @array-reply of details about all Redis commands.

Cluster clients must be aware of key positions in commands so commands can go to matching instances, but Redis commands vary between accepting one key, multiple keys, or even multiple keys separated by other data.

Starting from Redis 7.0.0 a more flexible mechanism was introduced in order to help clients finding the key positions.
For more information please check the [key-specs page][tr].
[tr]: /topics/key-specs

You can use `COMMAND` to cache a mapping between commands and key positions for
each command to enable exact routing of commands to cluster instances.

`COMMAND` has several subcommands:

>  - `COMMAND INFO`

Reviewer: COMMAND INFO was improved in 7.0 to be able to take 0 arguments and behave the same as COMMAND COMMAND.
i guess this PR should update command-info.md

Author: nothing in command-info.md says that COMMAND INFO takes at least one command + the recent change appears in the history section

I guess that's enough, no?

Reviewer: ok

 - `COMMAND COUNT`
 - `COMMAND GETKEYS`
 - `COMMAND LIST`

>  - `COMMAND DETAILS`

Reviewer: ```suggestion
 - `COMMAND DOCS`
```

`COMMAND`'s reply is an @array-reply, where each element is an @array-reply by itself, containing the following elements:

 - command name
 - command arity specification
 - nested @array-reply of command flags
 - position of first key in argument list
 - position of last key in argument list
 - step count for locating repeating keys
 - nested @array-reply of [ACL categories][ta]
 - nested @array-reply of [command tips][tb]
 - nested @array-reply of [key-specs][td]
 - nested @array-reply of subcommands
[ta]: /topics/acl
[tb]: /topics/command-tips

> [td]: /topics/key-specs

Reviewer: no sure about this markdown, is it sane?

Author: @itamarhaber ?

Reviewer 2: The syntax is ok, but there should be a \n before these just for safety (i.e. regular md is ok with that, but the current site's implementation is more sensitive).

Reviewer: @itamarhaber what do you mean? an empty line between the `- nested` list and the `[ta]` thing?
i must admit i don't understand any of this.. what does it do?

anyway, and you add a suggestion to fix it, or even just push a fix?

Reviewer 2: Yes, will push the fixes.

[tr]: /topics/key-specs

You can use `COMMAND` to cache a mapping between commands and key positions for
each command to enable exact routing of commands to cluster instances.

`COMMAND`'s reply is an @array-reply, where each element is an @array-reply by itself, containing the following elements:

 - command name
 - command arity specification
 - nested @array-reply of command flags
 - position of first key in argument list
 - position of last key in argument list
 - step count for locating repeating keys
 - nested @array-reply of ACL categories
 - nested @map-reply of additional information

## Example
```
1) 1) "mget"
   2) (integer) -2
   3) 1) readonly
      2) fast
   4) (integer) 1
   5) (integer) -1
   6) (integer) 1
   7) 1) @read
      2) @string
      3) @fast
   8)  1) "summary"
       2) "Get the values of all the given keys"
       3) "since"
       4) "1.0.0"
       5) "group"
       6) "string"
       7) "complexity"
       8) "O(N) where N is the number of keys to retrieve."
       9) "arguments"
      10) 1) 1) "name"
             2) "key"
             3) "type"
             4) "key"
             5) "flags"
             6) 1) multiple
             7) "value"

>              8) "key"

Reviewer: not sure if MGET is a good example, this is a bit complex, maybe find a simpler one)

Returns @array-reply of details about all Redis commands.

Cluster clients must be aware of key positions in commands so commands can go to matching instances, but Redis commands vary between accepting one key, multiple keys, or even multiple keys separated by other data.

Starting from Redis 7.0 a more flexible mechanism was introduced in order to help clients finding the key positions.
For more information please check the [key-specs page][tr].
[tr]: /topics/key-specs

You can use `COMMAND` to cache a mapping between commands and key positions for
each command to enable exact routing of commands to cluster instances.

`COMMAND` has several subcommands:
 - `COMMAND INFO`
 - `COMMAND COUNT`
 - `COMMAND GETKEYS`
 - `COMMAND LIST`

`COMMAND`'s reply is an @array-reply, where each element is an @array-reply by itself, containing the following elements:
 - command name
 - command arity specification
 - nested @array-reply of command flags
 - position of first key in argument list
 - position of last key in argument list
 - step count for locating repeating keys
 - nested @array-reply of ACL categories

>  - nested @map-reply of additional information

Reviewer 2: The current website doesn't support RESP3 return types, nor does it feature the RESP3 spec :/
```suggestion
 - nested @array-reply of additional information as a map
```

Examples:

  - `GET` arity is 2 since the command only accepts one argument and always has the format `GET _key_`.
  - `MGET` arity is -2 since the command accepts at a minimum one argument, but up to an unlimited number: `MGET _key1_ [key2] [key3] ...`.

Also note with `MGET`, the -1 value for "last key position" means the list
of keys may have unlimited length.

## flags

Command flags is @array-reply containing one or more status replies:
  - `write`: command may result in modifications
  - `readonly`: command will never modify keys
  - `denyoom`: reject command if currently out of memory
  - `admin`: server admin command
  - `pubsub`: pubsub-related command
  - `noscript`: deny this command from scripts
  - `random`: command has random results, dangerous for scripts

>   - `sort_for_script`: if called from script, sort output (Deprecated since 7.0.0)

Reviewer: let's either remove both, or refer to tips

>   - `loading`: allow command while database is loading

Reviewer: `no_async_loading` missing

Author: what about CMD_PROTCTED? do we want it here?
also, it's not documented next to the other CMD_*, just about the definition of `redisCommand`

Reviewer: IIRC we decided not to expose CMD_PROTECTED in COMMAND command.
but we did expose `no_async_loading`

Author: we do expose "protected" (see addReplyFlagsForCommand)

Reviewer: i must be drunk, or have been.
seems like i did that in https://github.com/redis/redis/pull/10015
i'm leaning towards removing it rather than documenting it.
WDYT?

Reviewer 2: If it ain't reported, it doesn't need to be documented

Author: ok, let's hide it from COMMAND (i'll add it to my TODO list)

Also note with `MGET`, the -1 value for "last key position" means the list
of keys may have unlimited length.

## flags
Command flags is @array-reply containing one or more status replies:

  - `write`: command may result in modifications
  - `readonly`: command will never modify keys
  - `denyoom`: reject command if currently out of memory
  - `admin`: server admin command
  - `pubsub`: pubsub-related command
  - `noscript`: deny this command from scripts
  - `random`: command has random results, dangerous for scripts
  - `sort_for_script`: if called from script, sort output
  - `loading`: allow command while database is loading
  - `stale`: allow command while replica has stale data
  - `skip_monitor`: do not show this command in MONITOR
  - `asking`: cluster related - accept even if importing
  - `fast`: command operates in constant or log(N) time. Used for latency monitoring.

>   - `movablekeys`: keys have no pre-determined position. You must discover keys yourself or use `key-specs` (starting from Redis 7.0).

Reviewer: we should state that it refers to firstkey,lastkey,step being incomplete, and the client needs to either use COMAND GETKEYS, or keys-specs (>7.0)

   4) (integer) 1
   5) (integer) 1
   6) (integer) 1
   ...
```

> Some Redis commands have no predetermined key locations. For those commands,
> flag `movablekeys` is added to the command flags @array-reply. Your Redis
> Cluster client needs to parse commands marked `movablekeys` to locate all relevant key positions.

Reviewer: ```suggestion
Some Redis commands have no predetermined key locations.
For those commands, the `movablekeys` flag is added to the command flags @array-reply,
which denotes that the `first key`, `last key`, `step` fields are insufficient to find all the keys,
and Cluster clients needs to user other measures which are described below to locate them.
```

maybe change the info in the list above to just point here rather than be an incomplete description.

> Partial list of commands currently requiring key location parsing:

Reviewer: ```suggestion
Here are a few examples of commands that are marked with `movablekeys`:
```

Reviewer: i now realize that he was trying to document a full list, because for each of them he provided the instructions of how to find the keys if you write your own parser.
so maybe we do need a complete list.
@itamarhaber WDYT? ("complete" lists are bound to get outdated and incomplete)

Author: @itamarhaber ^

   4) (integer) 1
   5) (integer) 1
   6) (integer) 1
   ...
```

> Some Redis commands have no predetermined key locations.

Reviewer: ```suggestion
Some Redis commands have no easy to fine / predetermined key locations.
```

For those commands, the `movablekeys` flag is added to the command flags @array-reply,
which denotes that the (`first-key`, `last-key`, `key-step`) fields are insufficient to find all the keys,
and Cluster clients needs to user other measures which are described below to locate them.

Here are a few examples of commands that are marked with `movablekeys`:
  - `SORT` - optional `STORE` key, optional `BY` weights, optional `GET` keys
  - `ZUNION` -  keys stop after `numkeys` count arguments
  - `ZUNIONSTORE` -  keys stop after `numkeys` count arguments
  - `ZINTER` - keys stop after `numkeys` count arguments
  - `ZINTERSTORE` -  keys stop after `numkeys` count arguments
  - `MIGRATE` - keys start after the `KEYS` keyword (if the second argument is en empty string)
  - `EVAL` - keys stop after `numkeys` count arguments
  - `EVALSHA` - keys stop after `numkeys` count arguments

Also see `COMMAND GETKEYS` for getting your Redis server tell you where keys
are in any given command.

> Starting from redis 7.0 clients can use the `key-specs` section in the additional information section in order to deduce key positions.

Reviewer 3: ```suggestion
Starting from Redis 7.0, clients can use the `key-specs` section in the additional information section in order to deduce key positions.
```

which denotes that the (`first-key`, `last-key`, `key-step`) fields are insufficient to find all the keys,
and Cluster clients needs to user other measures which are described below to locate them.

Here are a few examples of commands that are marked with `movablekeys`:
  - `SORT` - optional `STORE` key, optional `BY` weights, optional `GET` keys
  - `ZUNION` -  keys stop after `numkeys` count arguments
  - `ZUNIONSTORE` -  keys stop after `numkeys` count arguments
  - `ZINTER` - keys stop after `numkeys` count arguments
  - `ZINTERSTORE` -  keys stop after `numkeys` count arguments
  - `MIGRATE` - keys start after the `KEYS` keyword (if the second argument is en empty string)
  - `EVAL` - keys stop after `numkeys` count arguments
  - `EVALSHA` - keys stop after `numkeys` count arguments

Also see `COMMAND GETKEYS` for getting your Redis server tell you where keys
are in any given command.

Starting from redis 7.0 clients can use the `key-specs` section in the additional information section in order to deduce key positions.
The only two commands which still require using `COMMAND GETKEYS` are `SORT` and `MIGRATE` (assuming the client can parse and uses the `key-specs` section).

> For more information please check the [key-specs page][tr].

Reviewer 2: ```suggestion

For more information please check the [key-specs page][tr].
```

> [tr]: /topics/key-specs

Reviewer 2: ```suggestion
```

In the case of `MSET`, keys are every other position so the step value is 2.  Compare
with `MGET` above where the step value is just 1.

## acl categories

Available starting from Redis 6.0

For more information please check the [acl page][tr].
[tr]: /topics/acl

## additional information

Available starting from Redis 7.0

> The last element of each element in `COMMAND`s reply is a a map with the following fields:

Reviewer: i think that "last" is not good, we need to specify the exact indexes (also for other fields)

 - `summary`
 - `since`
 - `group`
 - `complexity`
 - `doc-flags`
 - `deprecated-since`
 - `replaced-by`
 - `history`
 - `hints`
 - `arguments`
 - `key-specs`
 - `subcommands`

Only `summary`, `since`, and `group` are mandatory, the rest may be absent.

### summary

Short command description

### since

Debut Redis version of the command

### group

Group of the command. Possible values:
 - generic
 - string
 - list
 - set
 - sorted-set
 - hash
 - pubsub
 - transactions
 - connection
 - server
 - scripting
 - hyperloglog
 - cluster
 - sentinel
 - geo
 - stream
 - bitmap
 - module

### complexity

A short explantion about the command's time complexity

### doc-flags

An @array-reply of flgas that are relevant for documentation purposes. Possible values:
 - deprecated: The command is deprecated
 - syscmd: System command, not meant to be executed by normal users

### deprecated-since

If deprecated, from which version?

### replaced-by

If deprecated, which command replaced it?

### history

An @array-reply, where each element is also an @array-reply with two elements:
1. The version when something changed about the command interface
2. A short description of the changes

### hints

> An @array-reply of hints that are meant to help clients/proxies know how to behave with this command

Reviewer: maybe we can put a link to a TBD placeholder topic page that we can create now?

### arguments

An @array-reply, where each element is a @map-reply describing a command argument.
For more information please check the [command-arguments page][tr].
[tr]: /topics/command-arguments

### key-specs

An @array-reply, where each element is a @map-reply describing a method to locate keys within the arguments.
For more information please check the [key-specs page][tr].
[tr]: /topics/key-specs

### subcommands

Some commands have subcommands (Example: `REWRITE` is a subcommand of `CONFIG`).
This is an @array-reply, with the same format and specification of `COMMAND`'s reply.

@return

> @array-reply: nested list of command details. Commands are returned

Reviewer: maybe we should move that whole page to command-info.md, and have the command.md refer to it and say it's an array with the same format..?

Reviewer: lets also document other sub-commands like GETKEYS, LIST, etc.

In the case of `MSET`, keys are every other position so the step value is 2.  Compare
with `MGET` above where the step value is just 1.

> ## ACL categories

Reviewer 2: ```suggestion
## ACL Categories
```

Available starting from Redis 6.0

> For more information please check the [ACL page][tr].
> [tr]: /topics/acl

Reviewer 2: ```suggestion
For more information please check the [ACL page][ta].
[ta]: /topics/acl
```

> ## additional information

Reviewer 2: ```suggestion
## Additional Information
```

Available starting from Redis 7.0

The element at index 7 (zero-based) of each element in `COMMAND`s reply is a a map with additional information.

> For the complete structure of that map please check the [command-info page][tr].
> [tr]: /topics/command-info

Reviewer 2: ```suggestion
For the complete structure of that map please check the [command-info page][tb].
[tb]: /topics/command-info
```
